src/kernel/thread.c

```diff
@@ -171,17 +171,11 @@ void doReplyTransfer(tcb_t *sender, tcb_t *receiver, cte_t *slot, bool_t grant)
     }
 
 #ifdef CONFIG_KERNEL_MCS
-    if (receiver->tcbSchedContext && isRunnable(receiver)) {
-        if ((refill_ready(receiver->tcbSchedContext) && refill_sufficient(receiver->tcbSchedContext, 0))) {
-            possibleSwitchTo(receiver);
-        } else {
-            if (validTimeoutHandler(receiver) && fault_type != seL4_Fault_Timeout) {
-                current_fault = seL4_Fault_Timeout_new(receiver->tcbSchedContext->scBadge);
-                handleTimeout(receiver);
-            } else {
-                postpone(receiver->tcbSchedContext);
-            }
-        }
+    if (isRunnable(receiver) && ensureSchedulable(receiver, fault_type != seL4_Fault_Timeout)) {
+        /* Presently we don't allow a thread to fault when replied to
```

Comment:
That's new. I had intended that faults should be able to propogate across replys so we can handle timeout faults in nested servers. What is the reasoning behind removing this?

Reply:
This doesn't prevent faults being propogated across replies, it prevents a thread that had time-out faulted to immediately time-out fault when the fault handler replies to the fault. The code that propogates time-out faults over replies is in `doReplyTransfer` I think.

Reply:
This doesn't remove any cases that would fault from the code it replaces.

Reply:
This comment could also be made... legible at all.

Reply:
ok so - if the timeout fault handlers don't correctly handle the fault what happens now? we should fault again, IMO

Reply:
Right, I wasn't sure what to do here. I'm also in favour of faulting again. Another suggestion was to use to allow the response to the fault to differentiate between resume with recurring fault or to resume and let the thread be postponed.

Reply:
It should mirror what happens on a page fault - if the fault isn't handled, the thread will continue to fault

Reply:
What do you do if you want to resume it but have it be postponed?

Reply:
any thread with a timeout fault handler faults when it has no time. There's no exceptions. You could remove the fault handler I guess.

Reply:
Okay, if that's the rule then `ensureSchedulable` shouldn't need the second argument. I'll rebase and remove it.

Reply:
err except round robin threads. although that's a bit confusing. it doesn't really matter what the impl does as setting a timeout fault for an RR thread is not really valid, so we can do what we like as long as we don't break the kernel